specs/networking/libp2p-standardization.md

```diff
@@ -135,24 +161,369 @@ number of shard subnets is defined via `SHARD_SUBNET_COUNT` and the shard
 Each Gossipsub
 [Message](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pb/rpc.proto#L17-L24)
 has a maximum size of 512KB (estimated from expected largest uncompressed block
-size).
+size). Clients SHOULD reject messages that are over this size limit.
 
 The `data` field of a Gossipsub `Message` is an SSZ-encoded object. For the `beacon_block` topic,
 this is a `beacon_block`. For the `beacon_attestation` topic, this is
 an `attestation`.
 
-## Eth-2 RPC
-
-#### Protocol Id: `/eth/serenity/beacon/rpc/1`
-
-The [RPC Interface](./rpc-interface.md) is specified in this repository.
-
 ## Discovery
 
 Discovery Version 5
 ([discv5](https://github.com/ethereum/devp2p/blob/master/discv5/discv5.md))
-will be used for discovery. This protocol uses a UDP transport and specifies
-its own encryption, ip-discovery and topic advertisement. Therefore, it has no
+is used for discovery. This protocol uses a UDP transport and specifies
+its own encryption, IP-discovery and topic advertisement. Therefore, it has no
 need to establish streams through `multistream-select`, rather, act
 as a standalone implementation that feeds discovered peers/topics (ENR-records) as
-`multiaddrs` into the libp2p service.
+`multiaddrs` into the libp2p service. The libp2p service subsequently forms
+connections and streams with discovered peers.
+
+## Eth-2 RPC
+
+*There is a specification for the [RPC Interface](./rpc-interface.md) in this
+repository which has implementations for libp2p and non-libp2p client. This
+section focuses solely on the libp2p specification and provides significant
+libp2p-specific modifications - as suggested by @Zah.*
+
+
+### Specification
+
+#### Protocol Segregation
+
+Each RPC message is segregated into it's own libp2p protocol id, which is a string of the form:
+
+```
+<ProtocolPrefix>/<MessageName>/<SchemaVersion>/<Encoding>
+```
+
+With:
+* **ProtocolPrefix** -- the RPC messages are grouped into families identified
+  by a shared LibP2P protocol name prefix. A conforming implementation is
+  expected to support either all messages appearing within a family or none of
+  them. In this case, we use `/eth/serenity/rpc`.
+* **MessageName** -- each RPC request is identified by a name consisting of
+English letters, digits and underscores (_).
+* **SchemaVersion** -- a semantic version consisting of one or more numbers
+  separated by dots (.). Each schema is versioned to facilitate backward and
+  forward-compatibility when possible.
```

Comment:
I agree entirely. This is an important discussion. We get some benefits when pushing things into multi-stream select, such as not requiring a method id in the RPC anymore. There is some discussion on the versioning here: https://github.com/ethereum/eth2.0-specs/pull/1158
I'm interested to know your thoughts on the above discussion.

Reply:
> It's unnecessary granularity: If we want, we can just increment the whole protocol's version, even if we just update a single message.

The granularity helps manage upgrades such that when specialized clients use only a part of the protocol, they need not be upgraded in some scenarios. I think this is a win, specially with the broadcast parts.

> I suspect that often the versions will not be independent. 

any such dependencies can be specified in the message that requires them.

I'm concerned about the use of semver here - it is nice for humans selecting a collection of versions to test together before using in application, but automating message selection in-protocol based on it seems like a can of worms in terms of subtle issues. This ties in with using SSZ for encoding - SSZ has no concept of optional fields really - no way of introducing purely additive changes - meaning that every time a request or response is changed, we effectively need to create a new incompatible protocol version.

Reply:
> no way of introducing purely additive changes

There is one type of upgrade that is supported by SSZ - it's adding a field at the end of a struct.

Otherwise, I agree that using a single number as a version is reasonable enough for both humans and machines.